File: core-java/src/main/java/org/spine3/Repository.java
/*
 * Copyright 2015, TeamDev Ltd. All rights reserved.
 *
 * Redistribution and use in source and/or binary forms, with or without
 * modification, must retain the above copyright notice and the following
 * disclaimer.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
package org.spine3;

import com.google.common.eventbus.Subscribe;
import com.google.protobuf.Message;
import org.spine3.base.CommandContext;
import org.spine3.base.EventRecord;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.List;

import static com.google.common.base.Preconditions.checkNotNull;

/**
 * Base interface for aggregate root repositories.
 *
 * @param <R> the type of the aggregated root
 * @param <I> the type of the aggregated root id
 * @param <C> the type of the command to create aggregate root instance
 *
 * @author Mikhail Melnik
 * @author Alexander Yevsyukov
 */
public interface Repository<I extends Message,
                            R extends AggregateRoot,
                            C extends Message> {

    /**
     * Stores the passed aggregate root.
     *
     * @param aggregateRoot an instance to store
     */
    void store(R aggregateRoot);

    /**
     * Loads the an aggregate by given id.
     *
     * @param aggregateId id of the aggregate to load
     * @return the loaded object
     */
    R load(I aggregateId);

    /**
     * Processes the command by dispatching it one of the repository methods or
     * to a method of an aggregate root.
     * <p>
     * If the passed command is for an aggregate, its instance is loaded by ID
     * obtained from the passed command.
     * <p>
     * For more details on writing aggregate commands read
     * <a href="http://github.com/SpineEventEngine/core/wiki/Writing-Aggregate-Commands">"Writing Aggregate Commands"</a>.
     *
     * @param command the command to dispatch
     * @param context context info of the command
     * @return a list of the event records
     * @throws InvocationTargetException if an exception occurs during command dispatching
     *
     * @see <a href="http://github.com/SpineEventEngine/core/wiki/Writing-Aggregate-Commands">Writing Aggregate Commands</a>
     *
     */
    List<EventRecord> dispatch(Message command, CommandContext context) throws InvocationTargetException;

    /**
     * Creates, initializes, and stores a new aggregated root.
     * <p>
     * The initial state of the aggregate root is taken from the creation command.
     *
     * @param command creation command
     * @param context creation command context
     * @return a list of the event records
     * @throws InvocationTargetException if an exception occurs during command handling
     */
    @Subscribe
    List<EventRecord> handleCreate(C command, CommandContext context) throws InvocationTargetException;


    /**
     * Helper class for wiring repositories into command processing.
     */
    @SuppressWarnings("UtilityClass")
    class Converter {

        private static final String DISPATCH_METHOD_NAME = "dispatch";

        /**
         * Returns the reference to the method {@link #dispatch(Message, CommandContext)} of the passed repository.
         * @param repository the repository instance to inspect
         * @return reference to the method
         */
        public static MessageSubscriber toMessageSubscriber(Repository repository) {
            checkNotNull(repository);

            try {
                Method method = repository.getClass().getMethod(DISPATCH_METHOD_NAME, Message.class, CommandContext.class);
                final MessageSubscriber result = new MessageSubscriber(repository, method);
                return result;
            } catch (NoSuchMethodException e) {
                //noinspection ProhibitedExceptionThrown // this exception cannot occur, otherwise it is a fatal error
                throw new Error(e);
            }
        }

        private Converter() {
        }
    }
}

File: buildSrc/src/main/kotlin/io/spine/internal/gradle/git/Repository.kt
/*
 * Copyright 2023, TeamDev. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Redistribution and use in source and/or binary forms, with or without
 * modification, must retain the above copyright notice and the following
 * disclaimer.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package io.spine.internal.gradle.git

import io.spine.internal.gradle.Cli
import io.spine.internal.gradle.fs.LazyTempPath

/**
 * Interacts with a real Git repository.
 *
 * Clones the repository with the provided SSH URL in a temporal folder. Provides
 * functionality to configure a user, checkout branches, commit changes and push them
 * to the remote repository.
 *
 * It is assumed that before using this class an appropriate SSH key that has
 * sufficient rights to perform described above operations was registered
 * in `ssh-agent`.
 *
 * NOTE: This class creates a temporal folder, so it holds resources. For the proper
 * release of resources please use the provided functionality inside a `use` block or
 * call the `close` method manually.
 */
class Repository private constructor(

    /**
     * The GitHub SSH URL to the underlying repository.
     */
    private val sshUrl: String,

    /**
     * Current user configuration.
     *
     * This configuration determines what ends up in author and committer fields of a commit.
     */
    private var user: UserInfo,

    /**
     * Currently checked out branch.
     */
    private var currentBranch: String

) : AutoCloseable {

    /**
     * Path to the temporal folder for a clone of the underlying repository.
     */
    val location = LazyTempPath("repoTemp")

    /**
     * Clones the repository with [the SSH url][sshUrl] into the [temporal folder][location].
     */
    private fun clone() {
        repoExecute("git", "clone", sshUrl, ".")
    }

    /**
     * Executes a command in the [location].
     */
    private fun repoExecute(vararg command: String): String =
        Cli(location.toFile()).execute(*command)

    /**
     * Checks out the branch by its name.
     */
    fun checkout(branch: String) {
        repoExecute("git", "checkout", branch)

        currentBranch = branch
    }

    /**
     * Configures the username and the email of the user.
     *
     * Overwrites `user.name` and `user.email` settings locally in [location] with
     * values from [user]. These settings determine what ends up in author and
     * committer fields of a commit.
     */
    fun configureUser(user: UserInfo) {
        repoExecute("git", "config", "user.name", user.name)
        repoExecute("git", "config", "user.email", user.email)

        this.user = user
    }

    /**
     * Stages all changes and commits with the provided message.
     */
    fun commitAllChanges(message: String) {
        stageAllChanges()
        commit(message)
    }

    private fun stageAllChanges() {
        repoExecute("git", "add", "--all")
    }

    private fun commit(message: String) {
        repoExecute(
            "git",
            "commit",
            "--allow-empty",
            "--message=${message}"
        )
    }

    /**
     * Pushes local repository to the remote.
     */
    fun push() {
        repoExecute("git", "push")
    }

    override fun close() {
        location.toFile().deleteRecursively()
    }

    companion object Factory {
        /**
         * Clones the repository with the provided SSH URL in a temporal folder.
         * Configures the username and the email of the Git user. See [configureUser]
         * documentation for more information. Performs checkout of the branch in
         * case it was passed. By default, [master][Branch.master] is checked out.
         *
         * @throws IllegalArgumentException if SSH URL is an empty string.
         */
        fun of(sshUrl: String, user: UserInfo, branch: String = Branch.master): Repository {
            require(sshUrl.isNotBlank()) { "SSH URL cannot be an empty string." }

            val repo = Repository(sshUrl, user, branch)
            repo.clone()
            repo.configureUser(user)

            if (branch != Branch.master) {
                repo.checkout(branch)
            }

            return repo
        }
    }
}
